### DISCUSSION TITLE
Need a guide on how to use without Composer

### DISCUSSION THREAD

**Discussion by Rubinskiy** [halaxa/json-machine#58] [Q&A] [answered]
It is mentioned in the guide that it's possible to use by only cloning the repository.
If anyone would kindly give a test script to demonstrate this possibility.

**Comment by halaxa** [comment, marked as answer]
It is not recommended, because there's no easy autoloading capability. But you're free to clone the repo, checkout the desired version, try to use the library and `require_once` every file with the class reported as missing. If really needed, an issue or PR might be made to discuss how to automate this for usages without composer autoloading.

Edit: Since version `1.1.0` there is a built-in optional autoloader for the use cases without composer. See README.

**Reply by Rubinskiy** [reply]
Ah, just as I thought. so I'm required to add a `require_once` reference to the main `JsonMachine.php` file in `/src` right?

**Reply by halaxa** [reply]
Yes, for the time being. As well as all other missing classes. But JsonMachine.php is an entrypoint which "tells" you all other missing ones.

**Comment by a-sync** [comment]
For anyone with a similar problem, here is a loader u can place next to the src folder:
```
<?php
/**
 * A simple PSR-4 spec auto loader to allow json-machine to function the same as if it were loaded via Composer
 *
 * To use this just include this file in your script and the JsonMachine namespace will be made available
 *
 * i.e. require_once('/path/to/json-machine/Autoloader.php');
 *
 * See: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader-examples.md
 *
 * @param string $class The fully-qualified class name.
 * @return void
 */
spl_autoload_register(function ($class) {

    // project-specific namespace prefix
    $prefix = 'JsonMachine\\';

    // base directory for the namespace prefix
    $base_dir = __DIR__ . DIRECTORY_SEPARATOR . 'src' . DIRECTORY_SEPARATOR;

    // does the class use the namespace prefix?
    $len = strlen($prefix);
    if (strncmp($prefix, $class, $len) !== 0) {
        // no, move to the next registered autoloader
        return;
    }

    // get the relative class name
    $relative_class = substr($class, $len);

    // replace the namespace prefix with the base directory, replace namespace
    // separators with directory separators in the relative class name, append
    // with .php
    $file = $base_dir . str_replace('\\', '/', $relative_class) . '.php';

    // if the file exists, require it
    if (file_exists($file)) {
        require $file;
    }
});

```

**Reply by halaxa** [reply]
Nice. PR?
